int32[] current_m1
int32[] current_m2